torch/csrc/jit/graph_executor.cpp

```
@@ -409,70 +412,117 @@ struct GraphExecutorImpl {
+              JIT_ASSERT(state[input] != State::MaybeDefined);
+            }
+            // hoist the nodes in the LinearBlock body to be before the linear block
+            for(auto it = body->nodes().begin(); it != body->nodes().end();) {
+              auto block_node = *it++;
+              block_node->moveBefore(n);
+            }
+
+            for(size_t i = 0; i < n->outputs().size(); ++i)
+              n->outputs().at(i)->replaceAllUsesWith(body->outputs().at(i));
+
+            it.destroyCurrent();
+          }
+        } break;
+        case prim::AutogradAdd: {
+          auto a = n->input(0);
+          auto b = n->input(1);
+          // if one is undefined, we can just drop the add
+          if(state[a] == State::Undefined) {
+            // Undef + b == b
+            n->output()->replaceAllUsesWith(b);
+            it.destroyCurrent();
+          } else if(state[b] == State::Undefined) {
+            // a + Undef == a
+            n->output()->replaceAllUsesWith(a);
+            it.destroyCurrent();
+          } else if(state[a] == State::Defined && state[b] == State::Defined) {
+            // when both are defined, we can use a normal, optimizable add instruction
+            WithInsertPoint guard(n);
+            Value* new_add = toVar(a) + toVar(b);
+            state[new_add] = State::Defined;
+            n->output()->replaceAllUsesWith(new_add);
+            it.destroyCurrent();
+          } else {
+            // otherwise we have conditionally-defined things, and we need
+            // to actually run an AutogradAdd which will guard for undefs
+            // so we leave the op as is
+            state[n->output()] = State::MaybeDefined;
+          }
+        } break;
+        case prim::Undefined: {
+          state[n->output()] = State::Undefined;
+        } break;
+        default:
+          for(auto o : n->outputs()) {
+            state[o] = State::MaybeDefined;
+          }
```

Assuming the argument in line 449 is correct, do we really need this default case? Maybe it's better to make it fail, so we don't get unexpected regressions in the future?

---

Yes, but only for trivial reasons: we want this pass to work on _any_ graph, including those that do not have GradOf and AutogradAdd nodes. That way it is always legal to run and we don't have to have bookkeeping to know what kind of graph we have.